ReplaySubject and buffer with debounce not emitting values after first subscription

The issue described below started with version 2.2.10 and continues with 2.2.11. It works with 2.2.7 through 2.2.9.

I have a test where a set a `ReplaySubject` like so:

```
locationPublisher = ReplaySubject.create()
locationPublisher.onNext(mockLocation)
locationPublisher.onComplete()

locationObservable = locationPublisher.share()
```

Then, I subscribe twice to the observable (I'm simplifying the test case just to demonstrate the bug):
```
        locationObservable
            .buffer(locationObservable.debounce(5, TimeUnit.SECONDS))
            .test()
            .assertValueCount(2)


        locationObservable
            .buffer(locationObservable.debounce(5, TimeUnit.SECONDS))
            .test()
            .assertValueCount(2)
```
When I run the above test, it fails on the second test subscription on RxJava 2.2.10 or 2.2.11:
> java.lang.AssertionError: Value counts differ; expected: 2 but was: 1 (latch = 0, values = 1, errors = 0, completions = 1)
Expected :2 
Actual   :1 (latch = 0, values = 1, errors = 0, completions = 1)

--- a/src/main/java/io/reactivex/internal/operators/flowable/FlowableRefCount.java
+++ b/src/main/java/io/reactivex/internal/operators/flowable/FlowableRefCount.java
@@ -115,14 +115,15 @@ public final class FlowableRefCount<T> extends Flowable<T> {
 
     void terminated(RefConnection rc) {
         synchronized (this) {
-            if (connection != null && connection == rc) {
-                connection = null;
+            if (connection == rc) {
                 if (rc.timer != null) {
                     rc.timer.dispose();
+                    rc.timer = null;
                 }
-            }
-            if (--rc.subscriberCount == 0) {
-                source.reset();
+                if (--rc.subscriberCount == 0) {
+                    connection = null;
+                    source.reset();
+                }
             }
         }
     }

--- a/src/main/java/io/reactivex/internal/operators/observable/ObservableRefCount.java
+++ b/src/main/java/io/reactivex/internal/operators/observable/ObservableRefCount.java
@@ -112,14 +112,15 @@ public final class ObservableRefCount<T> extends Observable<T> {
 
     void terminated(RefConnection rc) {
         synchronized (this) {
-            if (connection != null && connection == rc) {
-                connection = null;
+            if (connection == rc) {
                 if (rc.timer != null) {
                     rc.timer.dispose();
+                    rc.timer = null;
                 }
-            }
-            if (--rc.subscriberCount == 0) {
-                source.reset();
+                if (--rc.subscriberCount == 0) {
+                    connection = null;
+                    source.reset();
+                }
             }
         }
     }

--- a/src/test/java/io/reactivex/internal/operators/flowable/FlowableRefCountTest.java
+++ b/src/test/java/io/reactivex/internal/operators/flowable/FlowableRefCountTest.java
@@ -14,6 +14,7 @@
 package io.reactivex.internal.operators.flowable;
 
 import static org.junit.Assert.*;
+import static org.mockito.ArgumentMatchers.any;
 import static org.mockito.Mockito.*;
 
 import java.io.IOException;
@@ -1454,4 +1455,22 @@ public class FlowableRefCountTest extends RxJavaTest {
             .assertComplete();
         }
     }
-}
+
+    @Test
+    public void upstreamTerminationTriggersAnotherCancel() throws Exception {
+        ReplayProcessor<Integer> rp = ReplayProcessor.create();
+        rp.onNext(1);
+        rp.onComplete();
+
+        Flowable<Integer> shared = rp.share();
+
+        shared
+        .buffer(shared.debounce(5, TimeUnit.SECONDS))
+        .test()
+        .assertValueCount(2);
+
+        shared
+        .buffer(shared.debounce(5, TimeUnit.SECONDS))
+        .test()
+        .assertValueCount(2);
+    }}

--- a/src/test/java/io/reactivex/internal/operators/observable/ObservableRefCountTest.java
+++ b/src/test/java/io/reactivex/internal/operators/observable/ObservableRefCountTest.java
@@ -14,6 +14,7 @@
 package io.reactivex.internal.operators.observable;
 
 import static org.junit.Assert.*;
+import static org.mockito.ArgumentMatchers.any;
 import static org.mockito.Mockito.*;
 
 import java.io.IOException;
@@ -22,7 +23,7 @@ import java.util.*;
 import java.util.concurrent.*;
 import java.util.concurrent.atomic.*;
 
-import org.junit.Test;
+import org.junit.*;
 import org.mockito.InOrder;
 
 import io.reactivex.*;
@@ -1267,8 +1268,6 @@ public class ObservableRefCountTest {
         .publish()
         .refCount();
 
-        o.cancel(null);
-
         o.cancel(new RefConnection(o));
 
         RefConnection rc = new RefConnection(o);
@@ -1411,5 +1410,24 @@ public class ObservableRefCountTest {
             .assertNoErrors()
             .assertComplete();
         }
+    }
+
+    @Test
+    public void upstreamTerminationTriggersAnotherCancel() throws Exception {
+        ReplaySubject<Integer> rs = ReplaySubject.create();
+        rs.onNext(1);
+        rs.onComplete();
+
+        Observable<Integer> shared = rs.share();
+
+        shared
+        .buffer(shared.debounce(5, TimeUnit.SECONDS))
+        .test()
+        .assertValueCount(2);
+
+        shared
+        .buffer(shared.debounce(5, TimeUnit.SECONDS))
+        .test()
+        .assertValueCount(2);
     }
 }

